trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/standardfunctions/StandardFunctionHandler.java

```diff
@@ -409,6 +412,41 @@ private Map<String, IFunctionModelHandler> getFunctionModels() {
 		fill(map, "__builtin_sub_overflow_p", overapproximateGccOverflowCheck);
 		fill(map, "__builtin_mul_overflow_p", overapproximateGccOverflowCheck);
 
+		// Atomic operations https://en.cppreference.com/w/c/atomic
+		// Preprocessing leads to: https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html
+
+		fill(map, "__atomic_load", this::handleAtomicLoad);
+		fill(map, "__atomic_exchange", this::handleAtomicExchange);
+		fill(map, "__atomic_store", this::handleAtomicStore);
+
+		fill(map, "__atomic_load_n", this::handleAtomicLoadN);
+		fill(map, "__atomic_store_n", this::handleAtomicStoreN);
+		fill(map, "__atomic_exchange_n", this::handleAtomicExchangeN);
+
+		fill(map, "__atomic_fetch_add", this::handleAtomicFetchAdd);
+		fill(map, "__atomic_fetch_sub", this::handleAtomicFetchSub);
+		fill(map, "__atomic_fetch_and", this::handleAtomicFetchAnd);
+		fill(map, "__atomic_fetch_or", this::handleAtomicFetchOr);
+		fill(map, "__atomic_fetch_xor", this::handleAtomicFetchXor);
+
+		fill(map, "__atomic_test_and_set", this::handleAtomicTestAndSet);
+		fill(map, "__atomic_clear", this::handleAtomicClear);
+
+		fill(map, "__atomic_compare_exchange",
+				(main, node, loc, name) -> handleUnsupportedFunctionByOverapproximation(main, loc, name,
+						new CPrimitive(CPrimitives.BOOL)));

```

Comment:
You need to understand the [specification](https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html#index-_005f_005fatomic_005fcompare_005fexchange_005fn) first (which is rather complicated) 😄:
>This built-in function implements an atomic compare and exchange operation. This compares the contents of *ptr with the contents of *expected. If equal, the operation is a read-modify-write operation that writes desired into *ptr. If they are not equal, the operation is a read and the current contents of *ptr are written into *expected. weak is true for weak compare_exchange, which may fail spuriously, and false for the strong variation, which never fails spuriously. Many targets only offer the strong variation and ignore the parameter. When in doubt, use the strong variation.
> 
>If desired is written into *ptr then true is returned and memory is affected according to the memory order specified by success_memorder. There are no restrictions on what memory order can be used here.
>
>Otherwise, false is returned and memory is affected according to failure_memorder. This memory order cannot be __ATOMIC_RELEASE nor __ATOMIC_ACQ_REL. It also cannot be a stronger order than that specified by success_memorder.

The implementation itself should not be too difficult, but I did not want to spend too much time to understand the specificationand it does not seems as important as load/store/... to me.